src/software/backend/simulation/simulator.cpp

```diff
@@ -1,63 +1,97 @@
- *
- * See https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr for more info and
- * examples
- */
-struct FirmwareWorldDeleter
-{
-    void operator()(FirmwareWorld_t* firmware_world) const
-    {
-        app_firmware_world_destroy(firmware_world);
-    };
-};
-
-
 Simulator::Simulator(const World& world) : physics_world(world)
 {
     for (auto physics_robot : physics_world.getFriendlyPhysicsRobots())
     {
         auto simulator_robot = std::make_shared<SimulatorRobot>(physics_robot);
-        friendly_simulator_robots.emplace_back(simulator_robot);
+        std::unique_ptr<PrimitiveManager, PrimitiveManagerDeleter> primitive_manager(
+            app_primitive_manager_create(), PrimitiveManagerDeleter());
+        this->simulator_robots.emplace(simulator_robot, std::move(primitive_manager));
     }
     simulator_ball = std::make_shared<SimulatorBall>(physics_world.getPhysicsBall());
-}
 
-void Simulator::stepSimulation(const Duration& time_step)
-{
     auto firmware_robot = SimulatorRobotSingleton::createFirmwareRobot();
     auto firmware_ball  = SimulatorBallSingleton::createFirmwareBall();
+    // Make sure to release ownership from the local variables so the firmware_world
+    // can take ownership
     FirmwareWorld_t* firmware_world_raw =
-        app_firmware_world_create(firmware_robot.get(), firmware_ball.get());
-    auto firmware_world = std::unique_ptr<FirmwareWorld_t, FirmwareWorldDeleter>(
+        app_firmware_world_create(firmware_robot.release(), firmware_ball.release());
+    firmware_world = std::unique_ptr<FirmwareWorld_t, FirmwareWorldDeleter>(
         firmware_world_raw, FirmwareWorldDeleter());
+}
+
+void Simulator::stepSimulation(const Duration& time_step)
+{
     SimulatorBallSingleton::setSimulatorBall(simulator_ball);
-    for (const auto& simulator_robot : friendly_simulator_robots)
+    for (auto iter = simulator_robots.begin(); iter != simulator_robots.end(); iter++)
     {
+        auto simulator_robot    = iter->first;
+        auto& primitive_manager = iter->second;
         SimulatorRobotSingleton::setSimulatorRobot(simulator_robot);
-        // TODO: Simulate primitives
-        // app_main(firmware_world.get(), PRIMITIVE STUFF);
+        app_primitive_manager_runCurrentPrimitive(primitive_manager.get(),
+                                                  firmware_world.get());
     }
 
     physics_world.stepSimulation(time_step);
 }
 
 void Simulator::setPrimitives(ConstPrimitiveVectorPtr primitives)
 {
-    // TODO: Set primitives
-    // https://github.com/UBC-Thunderbots/Software/issues/768
+    if (primitives)
+    {
+        auto& primitive_vector = *primitives;
+        for (auto iter = simulator_robots.begin(); iter != simulator_robots.end(); iter++)
+        {
+            auto simulator_robot                        = iter->first;
+            PrimitiveManager* primitive_manager_raw_ptr = iter->second.get();
```

Comment:
Should we store the `PrimitiveManager` inside the `simulator_robot`?

Reply:
That is an option I considered.

I think controlling the primitives _might_ be more confusing that way however because of how the `SimulatorRobotSingleton` world with the `firmware_world`.

**Reason 1**
`SimulatorRobotSingleton::setSimulatorRobot(simulator_robot);` this call should stay outside the `SimulatorRobot` class otherwise we have to start doing `shared_from_this` which starts getting more complicated, unless we are fine doing that.

So if the `PrimitiveManager` was owned by a `SimulatorRobot` I imagine starting a new primitive / running a primitive would look roughly like

```
SimulatorRobotSingleton::setSimulatorRobot(simulator_robot);
simulator_robot->startPrimitive(firmware_world);
```
Which is shorter than I thought, but this assumes the primitive encoding can be done in the `SimulatorRobot` as well

**Reason 2**
The `SimulatorRobot` would only be able to store the `ConstPrimitveVectorPtr`, not a single `Primitive` directly, unless the encoding was done outside the class. I think it makes sense for the primitives and encoding to live in the same place, not split across classes.

Definitely would like some other thoughts @jared-paul  @jonathanlew @LiCody @akhilveeraghanta  on what is more clear.

Reply:
Hmm, so some thoughts on this:
- I think that at a high level, we can agree that the `PrimitiveManager` is _logically_ encapsulated in the robot? On the basis of that alone I think it's more clear.
- I would imagine the code looking more like:
```
SimulatorRobotSingleton::setSimulatorRobot(simulator_robot); // Renamed on personal preference, up to you
SimulatorRobotSingleton::runPrimitiveOnCurrentSimulatorRobot(primitive_index, primitive_params);
```
If I understand the implementation correctly, there is no real meaning to running a primitive on a robot if it's not in the singleton, so IMO there should be no way to do so.

Reply:
That's fair and I think your suggestion with the singleton is more clear. The only issue is that in order for the Singleton to be able to run the primitve on a SimulatorRobot, `runPrimitive` has to be in the public interface of the `SimulatorRobot` class, meaning it could potentially be called directly rather than via the singleton.

Reply:
> i would prefer not use use friend classes if possible, so we may just have to deal with it unless you have suggestions. 

Why not? Seems like an entirely appropriate use-case for the `friend` class.

Reply:
If I understand whats going on here I agree that this is an appropriate use-case for friends.

nit: Could we rename PrimitiveManager to "PrimitiveTracker" or something specific, it's hard to infer what it does through its name

Reply:
> nit: Could we rename PrimitiveManager to "PrimitiveTracker" or something specific, it's hard to infer what it does through its name

:arrow_double_up: for this

Reply:
Hmm. That's tricky. It's responsible for:
- starting and stopping primitives
- storing the state for the running primitive
- determining when to start/stop primitives (ex. if we pass the same primitive twice, we effectively keep running it). 

I'm not sure `PrimitiveTracker` is any more indicative (though I totally agree that `PrimitiveManager` is super generic). Thoughts?

Reply:
It seems like it’s violating the single responsibility principle. Do you think it would be better if we split it up into multiple modules?

Reply:
Hmm. Would you mind taking a look at `primitive_manager.{c|h}` and seeing if there are any logical divisions? I don't see any super easy wins, but I'm really not attached to the current structure, happy to change it.

Reply:
I have changed the `SimulatorRobot` to have all its members `protected` or `private`, and made the `SimulatorRobotSingleton` a `friend` class so that it must be controlled that way. `simulator.cpp` is more clean now.

I still need to convert all the `SimulatorRobotSingleton` test cases to use the firmware `app` layer now.

Reply:
> Would you mind taking a look at `primitive_manager.{c|h}` and seeing if there are any logical divisions? I don't see any super easy wins, but I'm really not attached to the current structure, happy to change it.

I've been researching this past week, and I think the Mediator pattern would be good. But since it's in C we aren't able to implement it because it depends on OOP. So I think it's fine the way it is

Reply:
> Mediator pattern

TIL, that's a new one. Anyhow, thanks for looking into this!